Avoid spoofing email senders

SegiNyn opened the pull request.

ThiefMaster commented:
Is it really necessary to change the APIs internally?  At least what I had in mind was simply changing the logic in the emails module to determine how to use From and Reply-to and only changing the form label in the various other places...

Also, I think we agreed to `"name via Indico" <noreply...>` in the From, and not using the site title in a very generic way..

SegiNyn commented:
> Is it really necessary to change the APIs internally? At least what I had in mind was simply changing the logic in the emails module to determine how to use From and Reply-to and only changing the form label in the various other places...

I saw that you already had a function _rewrite_sender that makes changes to the From address that's why I removed that and added a new function. And I thought it was better to make the changes before creating the EmailMessage. Also, to be sure where in the emails module did you have in mind? `do_send_email`?

SegiNyn commented:
The changes to the names to sender_address instead of from_address isn't actually necessary but it makes it obvious to developers as well not just the users as is the case if only the label is changed.

ThiefMaster commented on indico/core/notifications.py:
```diff
@@ -167,11 +169,14 @@ def make_email(to_list=None, cc_list=None, bcc_list=None, from_address=None, rep
     cc_list = {cc_list} if isinstance(cc_list, str) else cc_list
     bcc_list = {bcc_list} if isinstance(bcc_list, str) else bcc_list
     reply_address = {reply_address} if (isinstance(reply_address, str) and reply_address) else (reply_address or set())
+    from_address = _get_actual_sender_address(sender_address)
+    if sender_address and from_address != sender_address:
+        reply_address |= {sender_address}
```
I'm not sure if (even though the RFC allows it) using multiple reply-to addresses is a good idea.

I wonder whether we shouldn't clean up the API of `make_email` to be less ambiguous. One possibility would be to have only the `sender_address` arg, and completely get rid of the `reply_address` options - and then rely on the existing logic to determine whether the address goes in From or Reply-to.

Or alternatively, to make the PR less invasive, keep the API mostly as it is, but simply make an explicitly specified `reply_address` override the one that may otherwise be taken from the `from_address`.

I'll think about it and try around a bit on my dev instance...

SegiNyn replied:
I have no strong opinion because in our case we use the reply_address option in make_email since we were converting all from_address to reply_address. But there are some notification emails functions which are explicitly passing the reply_address address to the make_email function. So maybe your second option would be better

SegiNyn replied:
Hi @ThiefMaster, I just rebased this pr. Did you decide yet if I should revert the naming of `sender_address` back to `from_address`?

ThiefMaster commented on indico/core/emails.py:
```diff
@@ -66,15 +66,26 @@ def send_email_task(task, email, log_entry=None):
+    # rewrite sender address to the fallback, and try to keep the relevant part of the original one
+    # if we have a name we use the name (the address already goes into reply-to), otherwise we use
+    # the address which looks ugly but is (probably?) better than just using the site name
+    display_name = f'{orig_name or orig_address} (via {site_title})'
```
I wonder if the `{orig} (via {site_title})` format string should be configurable, in case someone does not want the (english) "via" there...

I was first thinking about `indico.conf`, but since this now gets called from `make_email`, we could even use proper i18n for it. I guess "via" is not really fitting for every single language after all...

cc @tomasr8

ThiefMaster commented:
Just FYI, I plan to merge this end of next week, shortly before we're going to put this in production at CERN.
